components/datetime/src/pattern/mod.rs

```diff
@@ -315,3 +316,208 @@ impl Default for CoarseHourCycle {
+                    ..
+                }) => Some(w),
+                _ => None,
+            })
+            .ok_or(Error::UnsupportedPluralPivot)?;
+
+        let mut plural_pattern = PluralPattern {
+            pivot_field: *pivot_field,
+            variants: Vec::new(),
+        };
+        plural_pattern.add_variant(plural_category, pattern);
+        Ok(plural_pattern)
+    }
+
+    pub fn pivot_field(&self) -> Week {
+        self.pivot_field
+    }
+
+    pub fn add_variant(&mut self, plural_category: PluralCategory, pattern: Pattern) {
+        self.variants.push((plural_category, pattern));
+    }
+
+    pub fn patterns_iter(&self) -> impl Iterator<Item = &Pattern> {
+        self.variants.iter().map(|(_, pattern)| pattern)
+    }
+
+    pub fn patterns_iter_mut(&mut self) -> impl Iterator<Item = &mut Pattern> {
+        self.variants.iter_mut().map(|(_, pattern)| pattern)
+    }
+
+    // Returns the pattern with given category, falling back to the pattern with the Other category if not present.
+    pub fn get(&self, category: PluralCategory) -> Option<&Pattern> {
+        self.variants
+            .iter()
+            .filter(|(c, _)| *c == category || *c == PluralCategory::Other)
+            .reduce(|a, b| if a.0 == PluralCategory::Other { b } else { a })
+            .map(|(_, v)| v)
+    }
+}
+
+#[derive(Debug, PartialEq, Clone)]
+#[cfg_attr(
+    feature = "provider_serde",
+    derive(serde::Serialize, serde::Deserialize)
+)]
+pub enum PatternPlurals {
+    #[cfg_attr(feature = "provider_serde", serde(rename = "plural_patterns"))]
+    MultipleVariants(PluralPattern),
+    #[cfg_attr(feature = "provider_serde", serde(rename = "pattern"))]
+    SinglePattern(Pattern),
+}
+
+enum PatternPluralsIterator<'a, T>
+where
+    T: Iterator<Item = &'a Pattern>,
+{
+    MultipleVariants(T),
+    SinglePattern(Option<&'a Pattern>),
+}
+
+impl<'a, T> Iterator for PatternPluralsIterator<'a, T>
+where
+    T: Iterator<Item = &'a Pattern>,
+{
+    type Item = &'a Pattern;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self {
+            PatternPluralsIterator::MultipleVariants(iter) => iter.next(),
+            PatternPluralsIterator::SinglePattern(opt_pattern) => opt_pattern.take(),
+        }
+    }
+}
+
+enum PatternPluralsMutIterator<'a, T>
+where
+    T: Iterator<Item = &'a mut Pattern>,
+{
+    MultipleVariants(T),
+    SinglePattern(Option<&'a mut Pattern>),
+}
+
+impl<'a, T> Iterator for PatternPluralsMutIterator<'a, T>
+where
+    T: Iterator<Item = &'a mut Pattern>,
+{
+    type Item = &'a mut Pattern;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self {
+            PatternPluralsMutIterator::MultipleVariants(iter) => iter.next(),
+            PatternPluralsMutIterator::SinglePattern(opt_pattern) => opt_pattern.take(),
+        }
+    }
+}
+
+impl PatternPlurals {
+    pub fn patterns_iter(&self) -> impl Iterator<Item = &Pattern> {
+        match self {
```

Comment:
suggestion: a way to do this with less code (no need to write your own iterator!) would be to return an `Either::Left(singlepattern.iter())`/etc. Check out https://github.com/unicode-org/icu4x/blob/5b1bf8820b3c3759c34c418c3abeab0b8e58c015/utils/zerovec/src/varzerovec/mod.rs#L283-L292 for an example of this

Reply:
Nice! Done